### PR TITLE
[move-compiler] Improved declaration handling after a parse error

### PR DESCRIPTION
**Pull request by awelc**
## Description 

This PR finesses parsing resilience for a specific situation which creates questionable user experience in the IDE. This specific case occurs when a valid sequence item in on the following line from parsing error. Consider the following example:
```
    public struct SomeStruct has drop, copy {}

    public fun bar(param: SomeStruct): SomeStruct {
        param
    }

    public fun test7(param: SomeStruct): SomeStruct {
        param.b
        param.bar()
    }
```
The `param.b` expression is invalid (as the programmer is still typing) but this should not be be the reason to not compile the following `param.bar()` incorrectly (and generate another error about wrong return value.

The following example shows a scenario that's even worse:
```
     let v =
     let _tmp1 = 42;
     let _tmp2 = _tmp1 * param;
     let _tmp3 = _tmp1 + param;
```
Here, when a programmer starts typing an incomplete (and unparsable) `let v = `, if we don't compile declaration of `_tmp1` correctly, it would lead to unnecessary cascading errors to appear due to `_tmp1` being undefined.

## Test plan 

Additional tests have been added. All tests (including the modified ones) must pass

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [x] CLI: Compiler may generate slightly different errors as a result of the compilation process being more permissive with respect to parsing errors
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

File: external-crates/move/crates/move-compiler/src/parser/syntax.rs [hunk 1]
```diff
@@ -1546,6 +1563,9 @@ fn parse_sequence(context: &mut Context) -> Result<Sequence, Box<Diagnostic>> {
                 seq.push(item);
                 last_semicolon_loc = Some(current_token_loc(context.tokens));
                 if let Err(diag) = consume_token(context.tokens, Tok::Semicolon) {
+                    if continue_sequence_after_error(context, diag.as_ref().clone()) {
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/parser/syntax.rs:1566]: I'll leave for @cgswords to review, but personally, I would love to see some exploration on using newline information. I think it feels potentially a more "general" solution, but who knows
e.g.
```
x.foo
//  ^
//  CURSOR HERE
x.bar();
```
just looking for let doesn't help here but seeing that the next token is on another line helps a lot.

File: external-crates/move/crates/move-compiler/src/parser/syntax.rs [hunk 2]
```diff
@@ -1503,6 +1503,45 @@ fn parse_sequence_item(context: &mut Context) -> Result<SequenceItem, Box<Diagno
     ))
 }
 
+// Checks if parsing of a sequence should continue after encountering an error.
+fn continue_sequence_after_error(
```
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/parser/syntax.rs:1507]: Nit: `should_continue_sequence_parsing` or similar?

File: external-crates/move/crates/move-compiler/tests/move_2024/ide_mode/seq_item_after_incomplete.exp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+error[E03009]: unbound variable
+   ┌─ tests/move_2024/ide_mode/seq_item_after_incomplete.move:24:9
+   │
+24 │         f                // parses correctly but without semicolon
+   │         ^ Unbound variable 'f'
+
+error[E01002]: unexpected token
+   ┌─ tests/move_2024/ide_mode/seq_item_after_incomplete.move:25:9
+   │
+25 │         let _tmp1 = 42;
```
Comment (cgswords) [inline external-crates/move/crates/move-compiler/tests/move_2024/ide_mode/seq_item_after_incomplete.exp:37]: Can we also suppress this error in IDE mode? Maybe as a follow-up PR?

**Review by cgswords — APPROVED** [review]
LGTM modulo some small nits.

File: external-crates/move/crates/move-compiler/src/parser/syntax.rs (hunk 2 above)
Comment (awelc) [reply external-crates/move/crates/move-compiler/src/parser/syntax.rs]: Changed to `should_continue_sequence_after_error`

File: external-crates/move/crates/move-compiler/src/parser/syntax.rs [hunk 4]
```diff
@@ -1503,6 +1503,45 @@ fn parse_sequence_item(context: &mut Context) -> Result<SequenceItem, Box<Diagno
...
+    context: &mut Context,
+    diag: Diagnostic,
+    last_token_preceded_by_eol: bool,
```
[the remark this answers is not in the document]
Comment (awelc) [reply external-crates/move/crates/move-compiler/src/parser/syntax.rs]: Yeah, this wasn't very smart - I simply slapped another parameter when updating the previous version without thinking too much. Thanks! And fixed!

**Pull request merged**